### PR TITLE
Format and Lint

### PR DESCRIPTION
**Pull request by domoritz**
* Use a standard format to avoid future merge conflicts
* Use recommended tslint where possible to be more standard
* Split into commits so you can add formatting to exiting PRs without merge conflicts

### CODE REVIEW

File: test/transform.test.ts [hunk 1]
```diff
@@ -6,24 +6,24 @@ import {TimeUnit} from '../src/timeunit';
...
         and: [
-          {not: {timeUnit: 'yearmonthdate', field: 'd', equal: {year: 2008}}},
-          {or: [{field: 'a', equal: 5}]}
-        ]
-      }
-    }]);
-    assert.equal(localLogger.warns[0], log.message.dayReplacedWithDate('yearmonthday'));
-  }));
+          {not: {timeUnit: 'yearmonthday' as TimeUnit, field: 'd', equal: {year: 2008}}},
+          {or: [{field: 'a', equal: 5}]},
```
Comment (kanitw) [inline test/transform.test.ts]: No trailing commas please

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -135,8 +135,6 @@ After adding a new example, make sure to run `yarn build:example <examplename>`
 
 - `test/` - Code for unit testing. `test`'s structure reflects `src`'s directory structure.
 For example, `test/compile/` tests files inside `src/compile/`.
-  - Note that we prepend `/* tslint:disable:quotemark */` to all files under `test/compile`
```
Comment (kanitw) [inline CONTRIBUTING.md]: Why removing this?  Are you removing double quotes from all tests?  (I'd rather keep the double quote so we can copy spec to editor to test easily.)

File: src/bin.ts [hunk 3]
```diff
@@ -58,14 +56,19 @@ export interface BinParams extends BaseBin {
...
-  extent?: number[];  // VgBinTransform uses a different extent so we need to pull this out.
+  extent?: number[]; // VgBinTransform uses a different extent so we need to pull this out.
 }
 
 export function binToString(bin: BinParams | boolean) {
   if (isBoolean(bin)) {
     return 'bin';
   }
-  return 'bin' + keys(bin).map(p => varName(`_${p}_${bin[p]}`)).join('');
+  return (
```
Comment (kanitw) [inline src/bin.ts]: Is this really prettier?

File: src/channel.ts [hunk 4]
```diff
@@ -147,12 +147,27 @@ export const SINGLE_DEF_CHANNELS: SingleDefChannel[] = flagKeys(SINGLE_DEF_CHANN
...
 // when running the schema generator
 // export type SingleDefChannel = typeof SINGLE_DEF_CHANNELS[0];
-export type SingleDefChannel = 'x' | 'y' | 'x2' | 'y2' |
-  'longitude' | 'latitude' | 'longitude2' | 'latitude2' |
-  'row' | 'column' |
-  'color' | 'fill' | 'stroke' |
-  'size' | 'shape' | 'opacity' |
-  'text' | 'tooltip' | 'href' | 'key';
+export type SingleDefChannel =
+  | 'x'
```
Comment (kanitw) [inline src/channel.ts]: This preceding `|` style is extremely terrible

File: src/compile/axis/assemble.ts [hunk 5]
```diff
@@ -9,7 +9,7 @@ import {AxisComponent, AxisComponentIndex} from './component';
 
 function assembleTitle(title: string | FieldDefBase<string>[], config: Config) {
   if (isArray(title)) {
-    return title.map(fieldDef => fieldDefTitle(fieldDef, config)).join(', ');
+    return title.map((fieldDef) => fieldDefTitle(fieldDef, config)).join(', ');
```
Comment (kanitw) [inline src/compile/axis/assemble.ts]: Can we turn off this parenthesis thing?

File: src/compile/baseconcat.ts [hunk 6]
```diff
@@ -76,12 +83,14 @@ export abstract class BaseConcatModel extends Model {
...
         ...(title ? {title} : {}),
         ...(style ? {style} : {}),
-        ...(layoutSizeEncodeEntry ? {
-          encode: {
-            update: layoutSizeEncodeEntry
-          }
-        } : {}),
-        ...child.assembleGroup()
+        ...(layoutSizeEncodeEntry
+          ? {
```
Comment (kanitw) [inline src/compile/baseconcat.ts]: This is terrible new line

File: src/compile/common.ts [hunk 7]
```diff
@@ -180,20 +216,19 @@ export function mergeTitleFieldDefs(f1: FieldDefBase<string>[], f2: FieldDefBase
...
+      value: mergeTitleFieldDefs(v1.value, v2.value),
     };
   } else if (!isArray(v1.value) && !isArray(v2.value)) {
     return {
       explicit: v1.explicit, // keep the old explicit
-      value: v1.value === v2.value ?
-        v1.value : // if title is the same just use one of them
-        v1.value + ', ' + v2.value // join title with comma if different
+      value:
+        v1.value === v2.value
```
Comment (kanitw) [inline src/compile/common.ts]: another terrible newline

File: CONTRIBUTING.md (hunk 2 above)
Comment (domoritz) [reply CONTRIBUTING.md]: We don't use this consistently (aka when I wanted to use it, I coudln't) and now that the tests are much faster, I feel this is less necessary.

Comment (kanitw) [reply CONTRIBUTING.md]: Why is it related to faster test?

File: src/axis.ts [hunk 8]
```diff
@@ -73,19 +70,16 @@ export interface Axis extends VgAxisBase, Guide {
...
-
-
 /**
  * A dictionary listing whether a certain axis property is applicable for only main axes or only grid axes.
  * (Properties not listed are applicable for both)
  */
 export const AXIS_PROPERTY_TYPE: {
-  // Using Mapped Type to declare type (https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types)
-  [k in keyof VgAxis]: 'main' | 'grid' | 'both'
+  [// Using Mapped Type to declare type (https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types)
```
Comment (kanitw) [inline src/axis.ts]: ????

File: src/channel.ts (hunk 4 above)
Comment (domoritz) [reply src/channel.ts]: Is there an option in prettier?

File: src/axis.ts (hunk 8 above)
Comment (domoritz) [reply src/axis.ts]: lol

File: CONTRIBUTING.md (hunk 2 above)
Comment (domoritz) [reply CONTRIBUTING.md]: http://bit.ly/js2json

**Comment by kanitw** [comment]
Wonder if we should include format to autorun somewhere.  Otherwise people will forget to run it anyway.  That said, that's additional 16.67s.

**Comment by kanitw** [comment]
https://basarat.gitbooks.io/typescript/docs/tools/husky.html ?

**Comment by domoritz** [comment]
I don't like precommit hooks that take longer than a few ms. It'll just cause frustration and hopefully you have auto-format enabled in your editor.

**Comment by kanitw** [comment]
> I don't like precommit hooks that take longer than a few ms. It'll just cause frustration and hopefully you have auto-format enabled in your editor.

Let's see how long does it take.  Maybe it touch only edited files then it might be fast.  
There is no prettier plugin for VSCode yet I think (not sure).

**Comment by domoritz** [comment]
https://github.com/prettier/prettier-vscode

**Pull request merged**